### PR TITLE
refactor Secret Manager plugin

### PR DESCRIPTION
**Pull request by sergiyvamz**
### Summary

AWSSecretManagerPlugin refactor.

### Description

- optimize workflows
- avoid re-fetching credentials if they just been fetched
- improved exception handling

### Additional Reviewers

@karenc-bq 
@congoamz 
@hsuamz 
@davecramer 
@crystall-bitquill

### CODE REVIEW

**Review by aaron-congo — APPROVED** [review]

**Review by karenc-bq — APPROVED** [review]

**Review by crystall-bitquill — APPROVED** [review]

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java [hunk 1]
```diff
@@ -130,47 +133,75 @@ public AWSSecretsManagerPlugin(
...
+    this.logger.logDebug("Login failed. SQLState=" + exception.getSQLState(), exception);
+    return SQLSTATE_ACCESS_ERROR.equals(exception.getSQLState());
+  }
+
+  private void applySecretToProperties(Properties properties) {
+    if (this.secret != null) {
+      properties.put("user", secret.getUsername());
+      properties.put("password", secret.getPassword());
     }
   }
```
Comment (davecramer) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:176]: Some comments here on the strategy would be useful for future readers

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java [hunk 2]
```diff
@@ -113,6 +115,7 @@ public AWSSecretsManagerPlugin(
     this.logger = logger;
     this.secretId = propertySet.getStringProperty(SECRET_ID_PROPERTY).getValue();
     this.region = Region.of(propertySet.getStringProperty(REGION_PROPERTY).getValue());
+    this.secretKey = Pair.of(secretId, region);
```
[the remark this answers is not in the document]
Comment (sergiyvamz) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:128]: Added necessary checks.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java [hunk 3]
```diff
@@ -40,6 +40,7 @@
 import com.fasterxml.jackson.annotation.JsonProperty;
 import com.fasterxml.jackson.core.JsonProcessingException;
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.mysql.cj.util.StringUtils;
```
Comment (davecramer) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:43]: hmm... we don't want to pull this into the other new repo. This is GPL

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java [hunk 4]
```diff
@@ -130,47 +142,104 @@ public AWSSecretsManagerPlugin(
 
   @Override
   public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
-    final Pair<String, Region> secretKey = Pair.of(secretId, region);
+
     final Properties properties = new Properties();
     properties.putAll(connectionUrl.getOriginalProperties());
-    Secret secret = SECRET_CACHE.get(secretKey);
+
+    boolean secretWasFetched = updateSecret(false);
```
Comment (davecramer) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:149]: I don't think we need to return a boolean here. Just return the secret

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java [hunk 5]
```diff
@@ -130,47 +142,104 @@ public AWSSecretsManagerPlugin(
...
-          throw new SQLException(ERROR_GET_SECRETS_FAILED);
+      applySecretToProperties(properties);
+      attemptToLogin(properties, connectionUrl);
+
+    } catch (SQLException exception) {
+      if (isLoginUnsuccessful(exception) && !secretWasFetched) {
+        // Login unsuccessful with cached credentials
+        // Try to re-fetch credentials and try again
+
+        secretWasFetched = updateSecret(true);
```
Comment (davecramer) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:160]: at this point if you get the secret here you can go ahead

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java [hunk 6]
```diff
@@ -130,47 +142,104 @@ public AWSSecretsManagerPlugin(
...
+   * Updates credentials to provided properties.
+   *
+   * @param properties Properties to store credentials.
+   */
+  private void applySecretToProperties(Properties properties) {
+    if (this.secret != null) {
+      /** Updated credentials are stored in properties. Other plugins in the plugin chain may
+      * change them if needed. Eventually, credentials will be used to open a new connection in
+      * {@link DefaultConnectionPlugin#openInitialConnection}
+      */
```
Comment (karenc-bq) [inline src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:195]:
```suggestion
      /* Updated credentials are stored in properties. Other plugins in the plugin chain may
       change them if needed. Eventually, credentials will be used to open a new connection in
       {@link DefaultConnectionPlugin#openInitialConnection}
      */
```
```suggestion
      /** Updated credentials are stored in properties. Other plugins in the plugin chain may
      * change them if needed. Eventually, credentials will be used to open a new connection in
      * {@link DefaultConnectionPlugin#openInitialConnection}
      */
```

**Review by karenc-bq — APPROVED** [review]

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java (hunk 4 above)
Comment (sergiyv-bitquill) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:149]: secretWasFetched helps to identify whether another fetch of credentials is needed when login attempt failed.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java (hunk 5 above)
Comment (sergiyv-bitquill) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:160]: If secretWasFetched  is false then the credentials hasn't been updated and it makes no sense to try to login again so we just throw original exception.

File: src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java (hunk 3 above)
Comment (sergiyvamz) [reply src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/AWSSecretsManagerPlugin.java:43]: Sure, we'll take that into account.

**Pull request merged**